consensus/state.go

```diff
@@ -225,32 +252,145 @@ func (cs *ConsensusState) getRoundState() *RoundState {
+}
+
 func (cs *ConsensusState) OnStop() {
-	// It's asynchronous so, there's not much to stop.
-	cs.BaseService.OnStop()
+	cs.QuitService.OnStop()
+}
+
+/*
+	The following three functions can be used to send messages into the consensus state
+		which may cause a state transition
+*/
+
+// May block on send if queue is full.
+func (cs *ConsensusState) AddVote(valIndex int, vote *types.Vote, peerKey string) (added bool, address []byte, err error) {
+	if peerKey == "" {
+		cs.internalMsgQueue <- msgInfo{&VoteMessage{valIndex, vote}, ""}
+	} else {
+		cs.peerMsgQueue <- msgInfo{&VoteMessage{valIndex, vote}, peerKey}
+	}
+
+	// TODO: wait for event?!
+	return false, nil, nil
+}
+
+func (cs *ConsensusState) SetProposal(proposal *types.Proposal, peerKey string) error {
+
+	if peerKey == "" {
+		cs.internalMsgQueue <- msgInfo{&ProposalMessage{proposal}, ""}
+	} else {
+		cs.peerMsgQueue <- msgInfo{&ProposalMessage{proposal}, peerKey}
+	}
+
+	// TODO: wait for event?!
+	return nil
+}
+
+func (cs *ConsensusState) AddProposalBlockPart(height, round int, part *types.Part, peerKey string) error {
+
+	if peerKey == "" {
+		cs.internalMsgQueue <- msgInfo{&BlockPartMessage{height, round, part}, ""}
+	} else {
+		cs.peerMsgQueue <- msgInfo{&BlockPartMessage{height, round, part}, peerKey}
+	}
+
+	// TODO: wait for event?!
+	return nil
+}
+
+func (cs *ConsensusState) SetProposalAndBlock(proposal *types.Proposal, block *types.Block, parts *types.PartSet, peerKey string) error {
+	cs.SetProposal(proposal, peerKey)
+	for i := 0; i < parts.Total(); i++ {
+		part := parts.GetPart(i)
+		cs.AddProposalBlockPart(cs.Height, cs.Round, part, peerKey)
+	}
+	return nil // TODO errors
+}
+
+//----------------------------------------------
+// internal functions for managing the state
+
+func (cs *ConsensusState) updateHeight(height int) {
+	cs.Height = height
+}
+
+func (cs *ConsensusState) updateRoundStep(round int, step RoundStepType) {
+	cs.Round = round
+	cs.Step = step
 }
 
 // EnterNewRound(height, 0) at cs.StartTime.
 func (cs *ConsensusState) scheduleRound0(height int) {
 	//log.Info("scheduleRound0", "now", time.Now(), "startTime", cs.StartTime)
 	sleepDuration := cs.StartTime.Sub(time.Now())
-	go func() {
-		if 0 < sleepDuration {
-			time.Sleep(sleepDuration)
-			// TODO: event?
+	cs.scheduleTimeout(sleepDuration, height, 0, 1)
+}
+
+// Attempt to schedule a timeout by sending timeoutInfo on the tickChan.
+// The timeoutRoutine is alwaya available to read from tickChan (it won't block).
+// The scheduling may fail if the timeoutRoutine has already scheduled a timeout for a later height/round/step.
+func (cs *ConsensusState) scheduleTimeout(duration time.Duration, height, round int, step RoundStepType) {
+	cs.tickChan <- timeoutInfo{duration, height, round, step}
+}
+
+// send a msg into the receiveRoutine regarding our own proposal, block part, or vote
+func (cs *ConsensusState) sendInternalMessage(mi msgInfo) {
```

Comment:
Instead of select { internal <-; <-timeout }, how about select { internal <-; default CList.PushBack(msg) } ?  We can shuffle more from CList to internalMsgQueue on the bottom of receiveMessage routine's for-loop.

We don't want to wait 10 milliseconds on things, and launching a new goroutine would make the order of sends to internalMsgQueue nondeterministic.